AWS deployment

manics opened the pull request:
~~This is a draft to demonstrate the full set of changes (it's currently deployed and JupyterHub desktops are working), I'll break things out into a separate PR.~~

~~Rebased on top of the K8TRE used at the York TREvolution demo.~~

Rebased on top of main

manics committed:
Install loadbalancer by vendor label

aws-load-balancer-controller

manics committed:
Add AWS cluster autoscaler

manics committed:
Add Gateway CRDs

vvcb commented on agnostics/crds/aws/base/kustomization.yaml:
@manics the Gateway API CRDs need to be installed before installing Cilium networking. Both should happen before K8TRE can be deployed. 

For Azure this is handled via Terraform. For K3S there are instructions here -> https://docs.k8tre.org/latest/guides/installation/

Everything before this step https://docs.k8tre.org/latest/guides/installation/#k8tre is the responsibility of the host organisation to provide before the cluster is ready to accept K8TRE.

@m1p1h can you confirm this please?

vvcb commented on appsets/agnostics/loadbalancer.yaml:
There is already an appset for `metallb`. For Azure, we just enable Azure Loadbalancer in terraform when setting up the cluster. If EKS does not support this, then it would make sense to move the `metallb` manifests into the `loadbalancer` folder. In hindsight, I should have named that folder `loadbalancer`!

Have you considered https://docs.aws.amazon.com/eks/latest/userguide/automode.html?

AaronJackson committed:
Samba Active Directory

Remove optional

Remove optional

Create keytabs in appropriate project namepsace

Create keytabs in appropriate project namepsace

ClusterRole for multiple namespaces

ClusterRole for multiple namespaces

VDI firewall to AD

VDI firewall to AD

VDI firewall to AD

VDI firewall to AD

VDI firewall to AD

manics committed:
Add AWS externaldns

manics committed:
Add AWS SSM Parameters ExternalSecrets

manics committed:
Add AWS SSM Parameterstore to ci/create-ci-secrets.py

manics committed:
Don't specify serviceAccountRef for External Secrets Operator

manics committed:
Add default CiliumClusterwideNetworkPolicy

manics committed:
Add placeholder AWS storage class readme

manics committed:
Add gateway/aws

manics committed:
Use self-signed cert for AWS Gateway for now

manics committed:
Disable ingress/cr8tor (replaced by HTTPRoute/cr8tor-approval-route)

manics committed:
eks-charts v0.0.226, aws-load-balancer-controller 3.1.0

manics committed:
Always install AD (not just for K3s)

use storage class rwo-default

manics committed:
Use pre-built pushed image ghcr.io/manics/k8tre-samba-dc:main

manics committed:
cilium-gateway-internal-gateway: instance instead of ip

manics committed:
https://github.com/kubernetes-sigs/gateway-api/releases/tag/v1.5.0

manics committed:
external-dns: add gateway-httproute gateway-tlsroute

manics committed:
external-dns: ServerSideApply=true

manics committed:
external-dns: restrict to private zone

manics committed:
gateway: add `load_balancing.cross_zone.enabled=true`

manics committed:
Add spec.infrastructure to gateway instead of manually patching in a job

manics committed:
Revert "Add Gateway CRDs"

This reverts commit a65be35e4a3133fcbefc1c0dd095a0ffc65303e3.
This reverts commit a2e43b992e4570943486214eeb4565dc1f8e7fb4.

manics committed:
external-dns ignore gateway-tlsroute

manics committed:
keycloak postgres storageClass: rwo-default

manics committed:
keycloak: debug logs

manics committed:
stg and prd: ghcr.io/k8tre/k8tre-backend:v1.1.0

manics committed:
aws k8tre-default-network-policy: "**.${DOMAIN}"

Add ENVIRONMENT DOMAIN to default-network-policy appset

manics committed:
Disable agnostics/default-network-policy.yaml for now

manics committed:
Reenable agnostics/default-network-policy.yaml otherwise nothing works

manics committed:
Add `host` to CiliumClusterwideNetworkPolicy/allow-pod-to-pod-via-gateway

manics committed:
Override some of project-namespace policy in aws

manics committed:
Try adding host remote-node to all

manics committed:
Try health unmanaged

manics committed:
Investigating possible hairpin routing node->lb->node

Disable client IP preservation

manics committed:
Revert AD

Revert "Use pre-built pushed image ghcr.io/manics/k8tre-samba-dc:main"

This reverts commit 86940674f8d0e010fee5ebb98b74daf4804269b4.

Revert "Always install AD (not just for K3s)"

This reverts commit 0c10bf55d78647e5cae6b56ca18ce4457b9bf31d.

Revert "Samba Active Directory"

This reverts commit 8340e4715050119d7cfe80416042bdb6e5272662.

manics committed:
loadbalancer: sync-wave: -10

manics committed:
Revert some NetworkPolicy changes from debugging connections problems

manics committed:
Set storageclass on JupyterHub db pvc

manics committed:
Lookup cluster name and region from labels

manics commented on agnostics/cluster-autoscaler/aws/base/values.yaml:
```diff
@@ -0,0 +1,13 @@
+autoDiscovery:
+  # TODO: it should be possible to template this in ArgoCD instead of hard-coding it
+  # https://artifacthub.io/packages/helm/cluster-autoscaler/cluster-autoscaler/9.51.0/#templating-the-autodiscovery-clustername
+  clusterName: k8tre-dev
```
The entire cluster name is substituted in aac7f1c51947f17c4d66681cbc8360e256ebd3c6

manics commented on agnostics/loadbalancer/aws/base/values.yaml:
```diff
@@ -0,0 +1,14 @@
+region: eu-west-2
+vpcTags:
+  Name: k8tre-dev
+
```
Done in aac7f1c51947f17c4d66681cbc8360e256ebd3c6

manics committed:
Remove duplicate unused appsets/agnostics/externaldns.yaml

manics committed:
create-ci-secrets.py: use ci-secrets.yaml throughout for consistency

manics committed:
create-ci-secrets.py: Use enum for backend type, make aws-ssm exceptions more specific

manics commented on ci/create-ci-secrets.py:
```diff
@@ -4,13 +4,14 @@
-External Secrets Operator in CI environments. It reads configuration from
+This script creates secrets in a backend store (Kubernetes Secrets or AWS Parameter Store)
+for use with External Secrets Operator in CI environments. It reads configuration from
 secrets.yaml and generates necessary passwords and keys automatically.
```
54afbce19fda6973919b2aba20901d58961a7131

manics commented on ci/create-ci-secrets.py:
```diff
@@ -442,16 +547,29 @@ def main(
-    
-    # Initialize secrets manager
+
+    # Select Backend
+    secrets_backend: SecretsBackend
+    if backend == "aws-ssm":
+        secrets_backend = AWSParameterStoreBackend(region=region, prefix=prefix)
+    else:
+        if not context:
+            console.print("[red]Error: --context is required for kubernetes backend[/red]")
+            raise typer.Exit(code=1)
+        secrets_backend = KubernetesBackend(context=context, namespace=namespace)
```
9361029dda678fd72c039639e8eb1b4fa7a65212

manics commented on ci/create-ci-secrets.py:
```diff
@@ -64,19 +78,174 @@ def generate_s3_access_key() -> str:
+           console.print(f"[red]Error checking parameter existence: {e}[/red]")
+           raise
+
+    def get_secret_data(self, secret_name: str) -> Dict[str, str]:
+        try:
+            response = self.ssm.get_parameter(Name=self._get_name(secret_name), WithDecryption=True)
+            return json.loads(response["Parameter"]["Value"])
+        except Exception:
+            return {}
```
https://github.com/k8tre/k8tre/commit/9361029dda678fd72c039639e8eb1b4fa7a65212

manics committed:
keycloak: KC_LOG_LEVEL=DEBUG for stg dev only

manics commented on apps/keycloak/base/values.yaml:
```diff
@@ -30,6 +30,8 @@ extraEnv: |
   - name: KC_HOSTNAME
     value: https://keycloak.${DOMAIN}
+  - name: KC_LOG_LEVEL
+    value: DEBUG
```
ab3f31cab1dc4bcdee0b8ba9a25e270eebe36640

manics commented on agnostics/cert-manager/aws/base/kustomization.yaml:
```diff
@@ -0,0 +1,8 @@
+resources:
+  - ../../base
+
+# TODO: enable ACME Lets Encrypt certificates, or use ACM on the load balancer
```
I'll deal with public certificates in a follow-up PR

manics commented on agnostics/default-network-policy/aws/base/clusterwide-networkpolicy.yaml:
```diff
@@ -0,0 +1,63 @@
+---
```
This is a quick fix to allow traffic to AWS services, I'll replace this with more fine-grained policies separately.

manics commented on agnostics/external-secrets/base/kustomization.yaml:
```diff
@@ -4,6 +4,10 @@ kind: Kustomization
+    valuesInline:
+      serviceAccount:
+        # Must match any external identity associations
+        name: external-secrets-sa
```
This needs to match https://github.com/k8tre/k8tre-aws/blob/ec064c2e85a7eb8d1f81502b59d372a70d8a4db4/k8tre-eks/serviceaccounts.tf#L108-L109 but it can be anything, so we can pick a name that works for Azure and AWS.

manics commented on agnostics/gateway/aws/base/gateway-infrastructure-patch.yaml:
```diff
@@ -0,0 +1,20 @@
+- op: add
+  path: /spec/infrastructure
+  value:
+    annotations:
```
Modern Cilium versions allow these annotations to be defined on the Gateway under `infrastructure` instead of [needing a job to patch the service](https://github.com/k8tre/k8tre/blob/5ab37ff97a81cfce9d1ee7a5589d3dc4979a3cc8/agnostics/gateway/azure/base/gateway-service-patch.yaml#L23-L42)

manics commented on agnostics/loadbalancer/aws/base/kustomization.yaml:
```diff
@@ -0,0 +1,17 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
We previously discussed moving this to the k8tre-aws repo, but since this requires installing CRDs I think it's easier to manage this with ArgoCD.

manics commented on agnostics/storage-class/aws/README.md:
```diff
@@ -0,0 +1,3 @@
+# AWS storage classes
+
+Storage classes `rwo-default` and `rwx-default` are created by https://github.com/k8tre/k8tre-aws
```
`rwx-default` uses EFS, which requires a fileSystemId. Since these classes are very simple it's easier to create them in k8tre-aws https://github.com/k8tre/k8tre-aws/blob/ec064c2e85a7eb8d1f81502b59d372a70d8a4db4/apps.tf#L43 instead of passing the fileSystemId to k8tre

Note I haven't defined any default storage class; there's no clear "best" class and since we have full control of Kubernetes/K8TRE we can ensure a storage class is always explicitly defined.

manics commented on apps/cr8tor/envs/dev/values.yaml:
```diff
@@ -80,18 +80,8 @@ publishService:
-  tls:
-    secretName: xk8tre-tls
+  # Using HTTPRoute/cr8tor-approval-route
+  enabled: false
```
Pretty sure we're not using ingress any more.

manics commented on apps/jupyterhub/base/values.yaml:
```diff
@@ -2,6 +2,9 @@ prePuller:
 hub:
+  db:
+    pvc:
+      storageClassName: rwo-default
```
This was missing from k8tre, which means `rwx-default` (typically NFS) would be used in Azure. It's best not to store a database on NFS.

manics commented on apps/keycloak/base/postgres.yaml:
```diff
@@ -9,6 +9,7 @@ spec:
   instances: 1
   storage:
     size: 512Mi # Reduced from 1Gi to use less storage
+    storageClass: rwo-default
```
It's best not to store a database on NFS. In the case of AWS EFS postgres refuses to start due to a mismatch in permissions.

manics commented on appsets/agnostics/default-network-policy.yaml:
```diff
@@ -1,7 +1,7 @@
 kind: ApplicationSet
 metadata:
-  name: externaldns
+  name: default-network-policy
```
This diff is misleading. I deleted `appsets/agnostics/externaldns.yaml` since it's unused (it duplicates `appsets/agnostics/external-dns.yaml`, but added `appsets/agnostics/default-network-policy.yaml`.

manics committed:
Fix context passing to create-ci-secrets.py log message

manics commented on ci/create-ci-secrets.py:
```diff
@@ -376,16 +476,17 @@ def print_summary(self, success: bool):
+                 console.print("\nTo verify the secrets, run:")
+                 console.print(
+                    f"[cyan]kubectl get secrets -n {self.backend.namespace} --context={self.backend.v1.api_client.configuration.host if hasattr(self.backend, 'v1') else 'context'}[/cyan]"
+                 )
```
1cf6e486ae6272122fe262a7714d87f498bd7193

manics committed:
Try replacing internal gateway cilium network load balancer with ALB ingress

- Ingress: host: `*.${DOMAIN}`
- Rely on ALBN to automatically look up matching certificate
- Add ingress to external-dns sources

vvcb commented on agnostics/loadbalancer/aws/base/values.yaml:
```diff
@@ -0,0 +1,14 @@
+vpcTags:
+  Name: ${CLUSTER_NAME}
+
+logLevel: debug
```
Best to set this only in the dev manifests if needed. Or better still to use a temporary overlay when troubleshooting (with argo sync turned off).

manics committed:
Fix typos and whitespace linting

manics committed:
AWS: use external-secrets not external-secrets-sa to match other clouds

manics committed:
Disable debug logging

manics commented:
I think I've fixed all issues highlighted by co-pilot
